modules/jpa/src/main/java/io/oasp/module/jpa/dataaccess/api/QueryDslHelper.java

```diff
@@ -0,0 +1,512 @@
+          return expression.isNotNull();
+        default:
+          throw new IllegalArgumentException("Operator " + operator + " does not accept null!");
+      }
+    } else if (v.isEmpty()) {
+      switch (operator) {
+        case LIKE:
+        case EQ:
+          return expression.isEmpty();
+        case NOT_LIKE:
+        case NE:
+          return expression.isNotEmpty();
+        default:
+          // continue;
+      }
+    }
+    StringExpression exp = expression;
+    if (ignoreCase) {
+      v = v.toUpperCase(Locale.US);
+      exp = exp.upper();
+    }
+    switch (operator) {
+      case LIKE:
+        return newLikeClause(exp, v, syntax, false, matchSubstring, false);
+      case NOT_LIKE:
+        return newLikeClause(exp, v, syntax, false, matchSubstring, true);
+      case EQ:
+        return exp.eq(v);
+      case NE:
+        return exp.ne(v);
+      case LT:
+        return exp.lt(v);
+      case LE:
+        return exp.loe(v);
+      case GT:
+        return exp.gt(v);
+      case GE:
+        return exp.goe(v);
+      default:
+        throw new IllegalCaseException(StringSearchOperator.class, operator);
+    }
+  }
+
+  /**
+   * @param query the {@link JPAQuery} to modify.
+   * @param expression the {@link StringExpression} to search on.
+   * @param value the string value or pattern to search for.
+   * @param config the {@link StringSearchConfigTo} to configure the search. May be {@code null} for regular equals
+   *        search.
+   */
+  protected void whereString(JPAQuery<?> query, StringExpression expression, String value,
+      StringSearchConfigTo config) {
+
+    BooleanExpression clause = newStringClause(expression, value, config);
+    if (clause != null) {
+      query.where(clause);
+    }
+  }
+
+  /**
+   * @param <T> generic type of the {@link Collection} values for the {@link SimpleExpression#in(Collection)
+   *        IN-clause}(s).
+   * @param expression the {@link SimpleExpression} used to create the {@link SimpleExpression#in(Collection)
+   *        IN-clause}(s).
+   * @param inValues the {@link Collection} of values for the {@link SimpleExpression#in(Collection) IN-clause}(s).
+   * @return the {@link BooleanExpression} for the {@link SimpleExpression#in(Collection) IN-clause}(s) oder
+   *         {@code null} if {@code inValues} is {@code null} or {@link Collection#isEmpty() empty}.
+   */
+  protected <T> BooleanExpression newInClause(SimpleExpression<T> expression, Collection<T> inValues) {
+
+    if ((inValues == null) || (inValues.isEmpty())) {
+      return null;
+    }
+    int size = inValues.size();
+    if (size == 1) {
+      return expression.eq(inValues.iterator().next());
+    }
+    int maxSizeOfInClause = DatabaseWorkaround.getInstance().getMaxSizeOfInClause();
+    if (size <= maxSizeOfInClause) {
+      return expression.in(inValues);
+    }
+    LOG.warn("Creating workaround for IN-clause with {} items - this can cause performance problems.", size);
```

Comment:
I am not sure, whether we should do this automatically.

* I think this should be opt-in. Not opt-out.
* The default `getInstance()` constructor even does not restrict anything as it will set maxSize to INT_MAX. 
* The `DatabaseWorkaround` class comes with a strange name, which does not correlate to the function of the class. It is just a value holder, not a workaround implementation.
* How can I adapt the value of DatabaseWorkaround maxSize? I think it is not possible besides calling the constructor in bootstrapping phase of my application. However, this is not documented anywhere and I think it is a bad design. We should go back and provide config properties for `maxInClauseSize` and/or `allowArbitraryLongInClauses`. However, as you can set the first property to MAX_INT, it is even up to discussion if you want to have the second property, which just makes sense if we have a good default value in mind besides MAX_INT.

Reply:
But this feature is already `opt-in` as it is disabled by default:
https://github.com/hohwille/oasp4j/blob/2fe1d69736fc62ed5e91ef5e6845587e8f7e654d/modules/jpa/src/main/java/io/oasp/module/jpa/dataaccess/api/DatabaseWorkaround.java#L29

In order to enable the feature you can subclass DatabaseWorkaround and instantiate it during bootstrap (e.g. via spring).

So all I am seeing valid here is that `DatabaseWorkaround` is an odd name. We should then agree on what would be a better name for it. A neutral name such as `DatabaseConfig`?
I would have used `application.properties` and use a custom spring property but that would IMHO not work well with the static access (and might make the solution unusable without spring-boot).

Reply:
> But this feature is already opt-in as it is disabled by default

ok, got it. I thought the "workaround" is to support a larger amount of IN clauses than a preconfigured maximum. So the "workaround" here is to restrict IN clauses manually to assure database misbehavior? This was not really clear for me by reading the javadocs.

> So all I am seeing valid here is that DatabaseWorkaround is an odd name. We should then agree on what would be a better name for it. A neutral name such as DatabaseConfig?

What about `SqlInClauseRestriction`? This is the only value, which is hold by this class.

> I would have used application.properties and use a custom spring property but that would IMHO not work well with the static access (and might make the solution unusable without spring-boot).

I do not get this entirely. You mean, as you don't have access to the config bean if also supporting non-spring frameworks with this? There are two a little bit ugly solutions:

1. Why not supporting both by configuring the class as SpringBean as well as general singleton? This is kind of a hack, but you could set the workaround beans instance on bootstrap when the bean is created and read the spring config at this time.
2. You could even just access the properties file in a static instance by the general properties interface of the jdk.

Both solutions do not provide any configuration at run-time, but at bootstrap time. However, I like them more to have the configuration in the general config in contrast to inherit some class to set one property. If we decide for this solution, we should also state in a comment in the properties, that to change this property, you have to restart the application.

Reply:
I see your point and agree. Will need some more effort and rework but will be worth it. Then also the class connected with the spring config can be package visible and then also the name is less important. Further you only have to set a property in `application.properties` to activate the feature.

For the problem itself: The IN clause bind variable design is entirely flawed. An IN clause should actually take one parameter which is a list of values such as primary keys of type Long. However, it is not like this. Instead for each value in the list a separate bind variable is created. This already leads to bad performance and spoofing of statement and execution plan caches. DBs also limit the number of values in an IN clause. Actually that is for reason and therefore we do not support any workaround for it by default. In most cases this is an indicator that you are doing something odd in your queries. Still there are project situations where you have to workaround this limitations. If that is the case, you have to write a lot of code. That is what is offered here only for those that actually need it. But if you ever run into this ugly situation you will love it. Otherwise just ignore it.